lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp

```diff
@@ -0,0 +1,235 @@
+
+  auto dispatched = xsmm::utils::buildXsmmCall(
+      rewriter, loc, dtype, dispatchOperands,
+      IntegerType::get(rewriter.getContext(), 64), module,
+      SymbolRefAttr::get(transposeOp->getContext(), dispatchName), nullptr);
+  rewriter.setInsertionPoint(transposeOp);
+  SmallVector<Value> operandRange{
+      getConstInt(rewriter, dyn_cast<DataTypeAttr>(dtype).getInt(), 64),
+      dispatched.getResult(0), input->getOperand(0), output->getOperand(1)};
+  auto invokeCall = xsmm::utils::buildXsmmCall(
+      rewriter, loc, dtype, operandRange, TypeRange(), module,
+      SymbolRefAttr::get(transposeOp->getContext(), invokeName), output);
+  return std::make_pair(&*dispatched, &*invokeCall);
+}
+
+static std::pair<Operation *, Operation *>
+convertTransposeOp(PatternRewriter &rewriter, Operation *transposeOp,
+                   Operation *input, Operation *output, Type outputType) {
+  LLVM_DEBUG(llvm::dbgs() << "BuildTransposeOp\n");
+  Value source = input->getOperand(0);
+  VectorType outType = cast<VectorType>(outputType);
+  if (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::TRANSPOSE,
+                                  outType)) {
+    memref::ExpandShapeOp expandShapeOp =
+        dyn_cast<memref::ExpandShapeOp>(source.getDefiningOp());
+    source = expandShapeOp.getSrc();
+    xsmm::UnaryInfo unaryInfo;
+    unaryInfo.m = expandShapeOp.getSrcType().getShape()[0];
+    unaryInfo.n = expandShapeOp.getSrcType().getShape()[1];
+    auto stridesOnInput = mlir::utils::getStaticStrides(source);
+    unaryInfo.ldi = stridesOnInput->front();
+    auto stridesOnOutput = mlir::utils::getStaticStrides(output->getOperand(1));
+
+    // Adjust ldo based on the VNNI factor.
+    unaryInfo.ldo =
+        stridesOnOutput->front() /
+        *vnni::utils::getVnniBlockingFactor(expandShapeOp.getSrcType());
+    auto functionOp = transposeOp->getParentOfType<func::FuncOp>();
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(&*functionOp.getBody().op_begin());
+
+    // If `OpTy` is unary or binary we need to dispatch and extra
+    // integer for the kind of operation to invoke.
+    return getTransposeOperation(rewriter, unaryInfo, outputType,
+                                 xsmm::UnaryKind::VNNI2, input, output,
+                                 transposeOp);
+  } else {
+    auto unaryInfo = xsmm::utils::getVectorUnaryInfo(
+        dyn_cast<MemRefType>(input->getOperand(0).getType()),
+        dyn_cast<MemRefType>(output->getOperand(1).getType()),
+        dyn_cast<VectorType>(input->getResult(0).getType()),
+        dyn_cast<VectorType>(output->getOperand(0).getType()),
+        xsmm::UnaryFlags::NONE);
+    auto temp = unaryInfo->n;
+    unaryInfo->n = unaryInfo->m;
+    unaryInfo->m = temp;
+
+    temp = unaryInfo->ldi;
+    unaryInfo->ldi = unaryInfo->ldo;
+    unaryInfo->ldo = temp;
+
+    return getTransposeOperation(rewriter, *unaryInfo, outputType,
```

Comment:
This is still duplicated code: `getTransposeOperation` is called with basically the same arguments on both sides of the branch.

As discussed before, use the if/else block to define the arguments then call the function only once.